File: algorithms/513. Find Bottom Left Tree Value/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findBottomLeftValue(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def dfs(node, height):
            if height > self.height:
                self.height = height
                self.target = node.val
            if node.left:
                dfs(node.left, height + 1)
            if node.right:
                dfs(node.right, height + 1)

        self.height = self.target = 0
        dfs(root, 1)
        return self.target

File: algorithms/513. Find Bottom Left Tree Value/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int findBottomLeftValue(TreeNode* root) {
        vector<TreeNode *> q;
        q.push_back(root);
        for ( int i = 0; i < q.size(); ++i ) {
            if ( q[i]->right ) {
                q.push_back(q[i]->right);
            }
            if ( q[i]->left ) {
                q.push_back(q[i]->left);
            }
        }
        return q.back()->val;
    }
};
